// ngx-gallery/lib/ngx-gallery-arrows.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, Output, EventEmitter, } from '@angular/core';
var NgxGalleryArrowsComponent = (function () {
    function NgxGalleryArrowsComponent() {
        this.onPrevClick = new EventEmitter();
        this.onNextClick = new EventEmitter();
    }
    NgxGalleryArrowsComponent.prototype.handlePrevClick = function () {
        this.onPrevClick.emit();
    };
    NgxGalleryArrowsComponent.prototype.handleNextClick = function () {
        this.onNextClick.emit();
    };
    return NgxGalleryArrowsComponent;
}());
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], NgxGalleryArrowsComponent.prototype, "prevDisabled", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], NgxGalleryArrowsComponent.prototype, "nextDisabled", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NgxGalleryArrowsComponent.prototype, "arrowPrevIcon", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NgxGalleryArrowsComponent.prototype, "arrowNextIcon", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NgxGalleryArrowsComponent.prototype, "onPrevClick", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NgxGalleryArrowsComponent.prototype, "onNextClick", void 0);
NgxGalleryArrowsComponent = __decorate([
    Component({
        selector: 'ngx-gallery-arrows',
        templateUrl: './ngx-gallery-arrows.component.html',
        styleUrls: ['./ngx-gallery-arrows.component.scss']
    })
], NgxGalleryArrowsComponent);
export { NgxGalleryArrowsComponent };
